substrate/frame/support/src/migrations.rs

```diff
@@ -344,3 +345,406 @@ impl<P: Get<&'static str>, DbWeight: Get<RuntimeDbWeight>> frame_support::traits
+	/// code.
+	InvalidCursor,
+	/// The migration encountered a permanent error and cannot continue.
+	Failed,
+}
+
+/// Notification handler for status updates regarding runtime upgrades.
+pub trait OnMigrationUpdate {
+	/// Notifies of the start of a runtime upgrade.
+	fn started() {}
+
+	/// Notifies of the completion of a runtime upgrade.
+	fn completed() {}
+
+	/// Infallibly handle a failed runtime upgrade.
+	///
+	/// Gets passed in the optional index of the migration that caused the failure.
+	fn failed(migration: Option<u32>) -> FailedUpgradeHandling;
+}
+
+/// How to proceed after a runtime upgrade failed.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum FailedUpgradeHandling {
+	/// Resume extrinsic processing of the chain. This will not resume the upgrade.
+	///
+	/// This should be supplemented with additional measures to ensure that the broken chain state
+	/// does not get further messed up by user extrinsics.
+	ForceUnstuck,
+	/// Do nothing and keep blocking extrinsics.
+	KeepStuck,
+}
+
+impl OnMigrationUpdate for () {
+	fn failed(_migration: Option<u32>) -> FailedUpgradeHandling {
+		FailedUpgradeHandling::KeepStuck
+	}
+}
+
+/// Something that can do multi step migrations.
+pub trait MultiStepMigrator {
+	/// Hint for whether [`Self::step`] should be called.
+	fn is_upgrading() -> bool;
+	/// Do the next step in the MBM process.
+	///
+	/// Must gracefully handle the case that it is currently not upgrading.
+	fn step() -> Weight;
+}
+
+impl MultiStepMigrator for () {
+	fn is_upgrading() -> bool {
+		false
+	}
+
+	fn step() -> Weight {
+		Weight::zero()
+	}
+}
+
+/// Multiple [`SteppedMigration`].
+pub trait SteppedMigrations {
+	fn len() -> u32;
+
+	fn nth_id(n: u32) -> Option<Vec<u8>>;
+
+	fn nth_step(
+		n: u32,
+		cursor: Option<Vec<u8>>,
+		meter: &mut WeightMeter,
+	) -> Option<Result<Option<Vec<u8>>, SteppedMigrationError>>;
+
+	fn nth_transactional_step(
+		n: u32,
+		cursor: Option<Vec<u8>>,
+		meter: &mut WeightMeter,
+	) -> Option<Result<Option<Vec<u8>>, SteppedMigrationError>>;
+
+	fn nth_max_steps(n: u32) -> Option<Option<u32>>;
+
+	fn cursor_max_encoded_len() -> usize;
+
+	fn identifier_max_encoded_len() -> usize;
+}
+
+impl<T: SteppedMigration> SteppedMigrations for T {
+	fn len() -> u32 {
+		1
+	}
+
+	fn nth_id(_: u32) -> Option<Vec<u8>> {
+		Some(T::id().encode())
+	}
+
+	fn nth_step(
+		_: u32,
+		cursor: Option<Vec<u8>>,
+		meter: &mut WeightMeter,
+	) -> Option<Result<Option<Vec<u8>>, SteppedMigrationError>> {
+		Some(
+			T::step(cursor.map(|c| Decode::decode(&mut &c[..]).unwrap()), meter)
+				.map(|v| v.map(|v| v.encode())),
+		)
+	}
+
+	fn nth_transactional_step(
+		_: u32,
+		cursor: Option<Vec<u8>>,
+		meter: &mut WeightMeter,
+	) -> Option<Result<Option<Vec<u8>>, SteppedMigrationError>> {
+		Some(
+			// FAIL-CI unwrap
+			T::transactional_step(cursor.map(|c| Decode::decode(&mut &c[..]).unwrap()), meter)
```

Comment:
is unwrap ok here?

Reply:
No its not, so i put the `FAIL-CI`. This makes the CI fail as a reminder, but thanks for commenting 😄

Reply:
fixed